src/Main.hs

```diff
@@ -236,27 +273,40 @@ main' settingsDirectory writeToken opts = do
   -- Users
   users <- readUsers settingsDirectory token (optsOffline opts)
 
-  -- Cache file
-  cachePath <- parseCachePath settingsDirectory org flow
+  flowsMap <- readFlows settingsDirectory token (optsOffline opts)
+  let flows = lookupFlows flowsMap org
+
+  case (flow `Prelude.elem` flows, searchAll) of
+    (False, False)  -> error $ "You are not a member of flow: " <> getParamName flow <> " in organisation: " <> getParamName org
+    (True , False)  -> doFlow token org ignoreCase by needle users showDate False flow
+    _               -> let needle' = (T.pack . getParamName $ flow) <> " " <> needle
+                       in mapM_ (doFlow token org ignoreCase by needle' users showDate True) flows
+  where
+    doFlow token org ignoreCase by needle users showDate showFlow aFlow = do
+      let maybeFlow = if showFlow then Just aFlow else Nothing
+
+      -- Cache file
+      cachePath <- parseCachePath settingsDirectory org aFlow
+
+      -- Read from cache
+      (rows, allRead) <- readRows cachePath `catch` onIOError ([], True)
+      lastRow <- grepRow users ignoreCase by needle showDate maybeFlow rows
+      when (not allRead) $ Prelude.putStrLn "Error: corrupted cache file, removing..." >> removeFile (toFilePath cachePath)
 
-  -- Read from cache
-  (rows, allRead) <- readRows cachePath `catch` onIOError ([], True)
-  lastRow <- grepRow users ignoreCase by needle rows
-  when (not allRead) $ Prelude.putStrLn "Error: corrupted cache file, removing..." >> removeFile (toFilePath cachePath)
+      -- Read from API
+      when (not $ optsOffline opts) $ do
+        mgr <- newManager tlsManagerSettings
+        onlineLoop mgr token org aFlow cachePath users ignoreCase by needle showDate maybeFlow lastRow
 
-  -- Read from API
-  when (not $ optsOffline opts) $ do
-    mgr <- newManager tlsManagerSettings
-    onlineLoop mgr token org flow cachePath users ignoreCase by needle lastRow
 
-onlineLoop :: Manager -> AuthToken -> ParamName Organisation -> ParamName Flow -> Path Abs File -> UserMap -> Bool -> Maybe Text -> Text -> Maybe Row -> IO ()
-onlineLoop mgr token org flow cachePath users ignoreCase by needle = go
+onlineLoop :: Manager -> AuthToken -> ParamName Organisation -> ParamName Flow -> Path Abs File -> UserMap -> Bool -> Maybe Text -> Text -> Bool -> Maybe (ParamName Flow) -> Maybe Row -> IO ()
+onlineLoop mgr token org flow cachePath users ignoreCase by needle showDate maybeFlow = go
   where go lastRow = do req <- untag <$> messagesRequest org flow (baseMessageOptions $ rowMessageId <$> lastRow)
                         let req' = authenticateRequest token req
                         res <- httpLbs req' mgr
                         rows <- mapMaybe messageToRow <$> throwDecode (responseBody res) :: IO [Row]
```

Comment:
`mapMaybe` seems to be ambigious given recent enough `unordered-containers`. Travis is unhappy.